### PR TITLE
Adjust highlights based on original text to remove extra spaces

### LINKED ISSUE #13: Reconstruct original text without extra whitespace tokens after highlighting
* Extra whitespace is added to text after tokenization and decoding from highlighting, we can remove the extra whitespace using the original text and update highlighted text characters accordingly

### PR DESCRIPTION
**Pull request by edwinzhng**
Fixes #13 

* Added a function to process highlighted text based on original text and adjust the highlight positions without extra spaces

* The algorithm is to greedily iterate using 2 pointers for both the original and highlighted text at the same time, where if characters are the same, then we can increment both pointers, otherwise if there is extra whitespace in one of the paragraphs, we increment the respective pointer

  * For example, if the highlighted text is "COVID - 19", and the original text is "COVID-19", then we iterate "C","C", -> "COVID","COVID" -> "COVID ","COVID" -> "COVID -", "COVID-"

  * This also handles the "[UNK]" token for characters that cannot be processed by the models

* Now we can just return the original paragraph directly!

### CODE REVIEW

File: api/app/services/highlighter.py [hunk 1]
```diff
@@ -65,8 +65,11 @@ def similarity_matrix(self, vector1, vector2):
         return (vector1.unsqueeze(1) * vector2.unsqueeze(0)).sum(-1)
 
     def highlight_paragraph(self, query_state, para_state,
-                            para_words) -> List[Tuple[int]]:
-        '''Returns the start and end positions of sentences that have the to'''
+                            para_words, original_paragraph) -> List[Tuple[int]]:
+        '''Returns the start and end character positions of highlighted sentences'''
+
+        if original_paragraph is None or original_paragraph == "":
```
Comment (rodrigonogueira4) [inline api/app/services/highlighter.py:71]:
```suggestion
        if original_paragraph:
```

This handles both `None` and `''`

File: api/app/services/highlighter.py [hunk 2]
```diff
@@ -102,7 +105,66 @@ def highlight_paragraph(self, query_state, para_state,
...
+        return self.adjust_highlights(original_paragraph, highlighted, highlights)
+
+    def adjust_highlights(self, original_text: str, highlighted_text: str,
+                          highlights: List[List[Tuple[int]]]):
+        """
+        Adjusts highlights based on extra spaces introduced by the tokenization process.
+        Iterates over highlighted and original text simultaneously to compute character positions.
+        """
+
+        if len(highlights) == 0:
```
Comment (rodrigonogueira4) [inline api/app/services/highlighter.py:119]:
```suggestion
        if highlights:
```
This handles both `None` and `[]`

File: api/app/services/highlighter.py [hunk 3]
```diff
@@ -102,7 +105,66 @@ def highlight_paragraph(self, query_state, para_state,
...
+            return original_text, []
```
Comment (rodrigonogueira4) [inline api/app/services/highlighter.py:120]: It should return only `[]`, right?

**Comment by edwinzhng** [comment]
> Thanks a lot for implementing this!
> 
> Could we use your method to find where to insert the highlight_token (lines 83-84) in the original text instead of using `para_words`? I'm asking that because the sentence tokenizer (lines 94-96) does not work well on tagged_paragraph. For example, the original text `The temperature is 24.5 celsius.` is tokenized into `The temperature is 24 . 5 celsius .` and, consequently, spacy's sentence tokenizer produces two sentences: `The temperature is 24.` and `5 celsius`.
> 
> Also, by inserting `highlight_token` in the original_text, we will not need `convert_tokens_to_strings` and `clean_up_tokenization` lines(86-92)

Hmm, I don't think we can reliably find the correct sentence endings with this algorithm... we would need to know exactly which "." characters represented the end of the sentence. For example, lots of of strings have a "." in abbreviations or author names, followed by a space, which is hard to parse. Maybe we should just ignore those cases for now?

**Comment by rodrigonogueira4** [comment]
> > Thanks a lot for implementing this!
> > Could we use your method to find where to insert the highlight_token (lines 83-84) in the original text instead of using `para_words`? I'm asking that because the sentence tokenizer (lines 94-96) does not work well on tagged_paragraph. For example, the original text `The temperature is 24.5 celsius.` is tokenized into `The temperature is 24 . 5 celsius .` and, consequently, spacy's sentence tokenizer produces two sentences: `The temperature is 24.` and `5 celsius`.
> > Also, by inserting `highlight_token` in the original_text, we will not need `convert_tokens_to_strings` and `clean_up_tokenization` lines(86-92)
> 
> Hmm, I don't think we can reliably find the correct sentence endings with this algorithm... we would need to know exactly which "." characters represented the end of the sentence. For example, lots of of strings have a "." in abbreviations or author names, followed by a space, which is hard to parse. Maybe we should just ignore those cases for now?

We would use spacy for sentence segmentation. The algorithm would require a small modification to what is already implemented:
```
# input
original_paragraph = "w1 w2 w3. w4 w5 w6."
para_words = ["w", "w2", "w3", ".", "w4", "w5", "w6", "."]

# Highlighting algorithm found w5 to be important.
highlighted_word = "w5"

# Mark the highlighted word with a special token "[HL]":
para_words = ["w", "w2", "w3", ".", "w4", "[HL]w5", "w6", "."]

# Mark in the original paragraph the highlighted word using Edwin's algorithm:
# That is, we iterate both sequences char by char, skiping whitespaces and 
# adding a special handling for the token [HL].
tagged_original_para = edwins_algo(original_paragraph, para_words)
# The output should be:
# tagged_original_para = "w1 w2 w3. w4 [HL]w5 w6."

```

Finally, we use spacy to segment the paragraph into sentences. In case the sentence contains [HL], remove it and return the sentence's  start and end positions.

**Comment by edwinzhng** [comment]
Ok, we now put highlights directly into the original text.

Results look good! Thanks for the suggestion @rodrigonogueira4 

image

File: api/app/services/highlighter.py [hunk 4]
```diff
@@ -91,9 +136,12 @@ def highlight_paragraph(self, query_state, para_state,
         tagged_paragraph = self.tokenizer.clean_up_tokenization(
```
Comment (rodrigonogueira4) [inline api/app/services/highlighter.py:136]: We don't need this anymore, right?

Comment (edwinzhng) [reply api/app/services/highlighter.py]: Fixed
